business_requirement/models/business.py

```diff
@@ -0,0 +1,348 @@
+        }
+    )
+    approval_date = fields.Datetime(
+        string='Approval Date',
+        copy=False,
+        readonly=True
+    )
+    approved_id = fields.Many2one(
+        'res.users',
+        string='Approved by',
+        copy=False,
+        readonly=True
+    )
+    company_id = fields.Many2one(
+        'res.company', string='Company',
+        required=True, readonly=True, states={'draft': [('readonly', False)]},
+        default=_get_default_company,
+    )
+    to_be_reviewed = fields.Boolean(
+        string='To be Reviewed'
+    )
+    kanban_state = fields.Selection(
+        [
+            ('normal', 'In Progress'),
+            ('on_hold', 'On Hold'),
+            ('done', 'Ready for next stage')
+        ],
+        'Kanban State',
+        track_visibility='onchange',
+        default='normal'
+    )
+    origin = fields.Char(
+        string='Source',
+        readonly=True,
+        states={
+            'draft': [('readonly', False)],
+            'confirmed': [('readonly', True)]
+        }
+    )
+    portal_published = fields.Boolean('In Portal', default=False)
+    user_id = fields.Many2one('res.users', string='Owner',
+                              default=lambda self: self.env.user,
+                              required=True)
+    date = fields.Date(
+        'Date',
+        default=lambda self: self._context.get(
+            'date', fields.Date.context_today(self)),
+        required=True
+    )
+
+    @api.model
+    def create(self, vals):
+        if vals.get('name', '/') == '/':
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'business.requirement')
+        return super().create(vals)
+
+    @api.multi
+    def write(self, vals):
+        if vals.get('state'):
+            br_xml_id = self.env.ref(
+                'business_requirement.group_business_requirement_manager')
+            user = self.env.user
+            grps = self.env.user.groups_id.ids
+            date = fields.Datetime.now()
+            if vals['state'] == 'confirmed':
+                vals.update({'confirmed_user_id': user.id,
+                             'confirmation_date': date})
+            if vals['state'] == 'draft':
+                vals.update({'confirmed_user_id': False,
+                             'approved_id': False,
+                             'confirmation_date': False,
+                             'approval_date': False
+                             })
+            if vals['state'] == 'approved':
+                if br_xml_id.id in grps:
+                    vals.update({'approved_id': user.id,
+                                 'approval_date': date})
+                else:
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed /cancel/drop.'))
+            if vals['state'] in ('approved', 'in_progress',
+                                 'done'):
```

Comment:
```suggestion
                                 'done'}:
```

Reply:
I suppose this is to make a set (so the opening one should be changed also) for improving the `in` operator efficiency. Isn't it?

Reply:
Yes. The previous comment it's for opening.